lib/collection/src/shards/shard_holder/mod.rs

```diff
@@ -581,6 +593,48 @@ impl ShardHolder {
         Ok(res)
     }
 
+    /// Common routing logic for reads and writes when using ShardKeyWithFallback
+    ///
+    /// If at least one of target shards is Active, use target shard. If not, redirect to fallback shard
+    pub fn route_with_fallback<'a>(
+        &self,
+        key: &'a ShardKeyWithFallback,
+    ) -> CollectionResult<(HashSet<ShardId>, &'a ShardKey)> {
+        let shard_key_to_ids_mapping = self.get_shard_key_to_ids_mapping();
+
+        let target_shard_ids = shard_key_to_ids_mapping.get(&key.target);
+        let fallback_shard_ids = shard_key_to_ids_mapping.get(&key.fallback);
+
+        if let Some(target_shard_ids) = target_shard_ids {
```

Comment:
I would try to make this function a little bit less nested. 
E.g.
`target_any_active` can account for both `target_shard_ids` being None and all shards being inactive